### ISSUE TITLE
Memory and redis issues on Worker instance

### ISSUE THREAD

**Issue by leobarcellos** [parcelvoy/platform#489]
Hey there! I'm experiencing two weird errors that I don't know what else to do anymore.

First one is about some sort of memory leak or anything like that:
`FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory`

<img width="526" alt="image" src="https://github.com/user-attachments/assets/a544f0ea-df29-47e6-b72c-fc9b9a1de6f2">

I already tried to increase memory of the instance, right now it's at 4GB and already tried to change the node max old space size using the `ENV NODE_OPTIONS="--max-old-space-size=3072"` env var. But it is still happening, not sure what else I could do to mitigate this.

Second one is related to Redis, I often receive a `ReplyError: READONLY You can't write against a read only replica.` -- My Redis instance is separated from the worker, but I'm not sure what is causing this. -- This one is happening frequently.

<img width="488" alt="image" src="https://github.com/user-attachments/assets/ffb5a6bd-3dde-4ae2-be83-afb35bc74bdc">
<img width="517" alt="image" src="https://github.com/user-attachments/assets/b62a82a5-2620-4b22-9f27-156e058509ff">

Any thoughts of how to handle it?

**Comment by leobarcellos** [comment]
I just ran these commands on redis, hope this will prevent the second error mentioned.

```
SLAVEOF NO ONE
REPLICAOF NO ONE
```

**Comment by pushchris** [comment]
I have not seen either of those before! How do you have redis setup? Are you in cluster mode? Using Elasticache or your own instance running?

**Comment by leobarcellos** [comment]
@pushchris I'm running on AWS ECS Fargate, I have a task with worker and a separated task with redis.
The "slave" error went away with the commands I've mentioned earlier.. however the first issue is the real problem here, I often get this memory heap error and it causes deadlock issues on databases etc

**Comment by leobarcellos** [comment]
Update: Redis issue was not related to memory issue.

I just updated my deploy to a EC2 "monolith" (ui / api / worker / redis) with AWS RDS

It's running smoothier than the ECS, however I just experienced the same error:

![Screenshot 2024-09-04 at 13 01 02](https://github.com/user-attachments/assets/273b9520-44cd-4d0c-bbd1-eac66d20781a)

@pushchris We have like 10 projects and each of them have a lot of lists, I've observed that all lists get updated at the same time once I open the project on the UI (is this the normal behavior ?) -- Also, everytime I get this heap memory error, I have like 2~4 campaigns running, sending 1k-2k emails each. 

Another info that might be useful, not sure if this is OK, but when I check worker logs, I get like SEVERAL log of queue:job:started, like really SEVERAL, screenshot is just illustrative because it prints like a lot.

![Screenshot 2024-09-04 at 13 09 06](https://github.com/user-attachments/assets/e1355373-4f44-405f-b033-886ef13f0e94)

Is this normal? Or something is wrong here?

**Comment by leobarcellos** [comment]
By the way, I just checked this answer on stack overflow:
https://stackoverflow.com/questions/55613789/how-to-fix-fatal-error-ineffective-mark-compacts-near-heap-limit-allocation-fa

I just updated the env for the worker like this: `NODE_OPTIONS=--max_old_space_size=7580`

Let's see if this helps.

**Comment by leobarcellos** [comment]
Well, it keep happening, however I'm not experiencing deadlock issues or stuck campaigns anymore. I guess it was happening only on ECS FARGATE.

![Screenshot 2024-09-05 at 13 32 47](https://github.com/user-attachments/assets/c0face3f-1be9-4a14-b6ef-42289b3e3dcd)

It's throwing errors but at least its working. Not sure what else I could do to fix this.

**Comment by leobarcellos** [comment]
Yeah, I was at `debug` level to actually see what's going on. I returned it back to `error` log level, but it's still happening.

I will drop the screenshot here, it oftens causes a deadlock error on journey_process job

Same errors, different days (actually it's happening every day) -- However it seems to be working just fine, besides the errors.

![Screenshot 2024-09-06 at 12 45 32](https://github.com/user-attachments/assets/f065d701-c4f5-4b7e-8184-b7cca15366b3)
![Screenshot 2024-09-09 at 13 00 05](https://github.com/user-attachments/assets/db868369-3199-467a-a6a6-027b64b944d2)

**Comment by leobarcellos** [comment]
More info, memory of worker instance keep increasing.. that's why things start to fail at certain point:

![Screenshot 2024-09-12 at 14 06 41](https://github.com/user-attachments/assets/51a8286a-cfc1-4ca4-89b2-4b7ab3d86bf4)

It starts like 50mb.. 70mb, and then it keeps increasing.. 300mb.. 700mb.. 1gb.. until maximum available.

I just changed docker compose to keep worker at 2GB maximum memory and I'm spawning 5 workers using `docker compose up worker --scale worker=5 -d`. Hopefully it will contain the problem.

However I don't know exactly where this memory leak is happening.
